### ISSUE TITLE
Editor layout is called 5 times opening 1 file

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#112630]
Steps to Reproduce:

1. add a breakpoint or `console.log` into `editorGroupView.layout`
2. run vscode and have no file open
3. click on a file from the explorer

=> `layout` is called 4 times

First one seems legit to me:

```
editorGroupView.js:1298 Error
    at EditorGroupView.layout (editorGroupView.js:1298)
    at LeafNode.layout (gridview.js:513)
    at VerticalViewItem.layout (splitview.js:76)
    at SplitView.layoutViews (splitview.js:642)
    at SplitView.relayout (splitview.js:549)
    at SplitView.onViewChange (splitview.js:414)
    at splitview.js:472
    at event.js:67
    at Emitter.fire (event.js:456)
    at event.js:44
    at Emitter.fire (event.js:456)
    at Emitter.fire (event.js:459)
    at EditorControl.doSetActiveEditorPane (editorControl.js:107)
```

But not sure about these:

```
Error
    at EditorGroupView.layout (editorGroupView.js:1298)
    at LeafNode.layout (gridview.js:513)
    at VerticalViewItem.layout (splitview.js:76)
    at SplitView.layoutViews (splitview.js:642)
    at SplitView.layout (splitview.js:299)
    at BranchNode.layout (gridview.js:216)
    at GridView.layout (gridview.js:619)
    at SerializableGrid.layout (grid.js:172)
    at SerializableGrid.layout (grid.js:437)
    at GridWidgetView.layout (editorPart.js:44)
    at CenteredViewLayout.layout (centeredViewLayout.js:74)
    at EditorPart.doLayout (editorPart.js:810)
    at EditorPart.layout (editorPart.js:805)
    at LeafNode.layout (gridview.js:513)
    at VerticalViewItem.layout (splitview.js:76)
    at SplitView.layoutViews (splitview.js:642)
    at SplitView.layout (splitview.js:299)
    at BranchNode.layout (gridview.js:216)
    at HorizontalViewItem.layout (splitview.js:76)
    at SplitView.layoutViews (splitview.js:642)
    at SplitView.relayout (splitview.js:549)
    at SplitView.onViewChange (splitview.js:414)
```

Also adding Isi since this is maybe the centered grid layout.

**Comment by bpasero** [comment]
This is bad for the work to support vertical tabs in https://github.com/microsoft/vscode/pull/106448 because the `layout` will be expensive when tabs wrap.

**Comment by joaomoreno** [comment]
I remember spending a **lot** of time making sure this wouldn't happen, as far as the grid goes. Sucks if it broke. I can take a peek.

**Comment by joaomoreno** [comment]
I actually see 6 calls to layout, this is preposterous.

**Comment by joaomoreno** [comment]
Those last 2 `layout` calls from `relayout` are:

- `TabsTitleControl.doLayout` @bpasero https://github.com/microsoft/vscode/blob/691951c3b1c742c2fd11490fe26086ae03c4d71e/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1317-L1317
- `TabsTitleControl.updateBreadcrumbsControl` despite nothing changes regarding breadcrumbs @jrieken https://github.com/microsoft/vscode/blob/691951c3b1c742c2fd11490fe26086ae03c4d71e/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L204-L204

**Comment by bpasero** [comment]
@joaomoreno I agree with reducing calls to `relayout` as much as possible. `relayout` should only be called if the title control knows for a fact that something has changed that impacts the entire splitview. 

I am not sure about the `doLayout` reference you pointed me to. That is just a way to schedule repeated layout calls on the next animation frame and is meant to reduce perf overhead of this method. It is the one calling this method (`layout`) that should not be called more than once.

**Comment by joaomoreno** [comment]
@bpasero This is the stack of that call:

```
 Error
    at EditorGroupView.layout (file:///home/joao/Work/vscode/out/vs/workbench/browser/parts/editor/editorGroupView.js:1298:35)
    at EditorGroupView.relayout (file:///home/joao/Work/vscode/out/vs/workbench/browser/parts/editor/editorGroupView.js:1313:22)
    at TabsTitleControl.doLayout (file:///home/joao/Work/vscode/out/vs/workbench/browser/parts/editor/tabsTitleControl.js:1041:28)
    at AnimationFrameQueueItem._runner (file:///home/joao/Work/vscode/out/vs/workbench/browser/parts/editor/tabsTitleControl.js:1011:26)
    at AnimationFrameQueueItem.execute (file:///home/joao/Work/vscode/out/vs/base/browser/dom.js:147:22)
    at animationFrameRunner (file:///home/joao/Work/vscode/out/vs/base/browser/dom.js:183:21)
```

Which runs after all other 5 calls run, so it seems it's just unnecessary, given the grid within has already been laid out correctly.

**Comment by joaomoreno** [comment]
Repro case:

1. Have no editors open
2. Open a text editor

🐛 6 calls to `layout` appear!

---

There are 3 separate causes to the 6 `layout` calls:

1. **Grid:** I found that the grid is indeed calling too many `layout` calls into its views. This must've crept up over time and can be fixed by having some sort of `SplitViewOptions.manualLayout?: boolean` option. Once that is fixed, we will reduce 6 to 4 calls. @joaomoreno 
2. **Grid within grid:** The editor group node within the workbench grid first reports having changed its size, due to it being about to open an editor, which causes a `layout`. Then, the editor node within the editor grid reports having changed its size, due to it being about to open a text editor; this then propagate to the outer workbench grid causing a second `layout`. This really only happen when going from 0 to 1 editor in the group. This is hard to fix.
3. **Combination with breadcrumbs and wrapping tabs:** There is code in `TabsTitleControl` to measure the title bar height due to the new wrapping tabs feature. Unfortunately there is also code in there to relayout whenever breadcrumbs appear and disappear. The combination of these two code paths cause two additiopnal `layout` calls, despite neither the breadcrumbs state nor wrapping tabs height actually change in 0 to 1 editor case. @bpasero

**Comment by bpasero** [comment]
Thanks, I will look into the `relayout` issues.

**Comment by jrieken** [comment]
> Unfortunately there is also code in there to relayout whenever breadcrumbs appear and disappear

@joaomoreno Do you mean when breadcrumbs are enabled/disabled or also when their content changes? In the case of the former we might assume that breadcrumbs are enabled. Surely this code was written when they were off by default.


Out of curiosity: it is always good to prevent unneeded calls, but do have a chance to protect yourself from such calls, e.g early return when layout is called with the input? I feel without that you are repeatedly encountering this because stuff just rots

**Comment by joaomoreno** [comment]
> @joaomoreno Do you mean when breadcrumbs are enabled/disabled or also when their content changes? In the case of the former we might assume that breadcrumbs are enabled. Surely this code was written when they were off by default.

Yeah that makes sense. What I observed was that the breadcrumbs code was assuming something has changed, even though nothing changes regarding breadcrumbs when going from 0 to 1 editor.

> Out of curiosity: it is always good to prevent unneeded calls, but do have a chance to protect yourself from such calls, e.g early return when layout is called with the input? I feel without that you are repeatedly encountering this because stuff just rots

Totally. This occurs now due to rot. Yeah I'll see if I can protect a bit better here.

**Comment by jrieken** [comment]
> What I observed was that the breadcrumbs code was assuming something has changed, even though nothing changes regarding breadcrumbs when going from 0 to 1 editor.

Breadcrumbs does that because per editor input it might be hidden, e.g when you open an extension editor it needs to hide, and that needs to relayout to return the screen real-estate (or take it back). Tho, surely can be done smarter

**Comment by bpasero** [comment]
@jrieken I think `breadcrumbsControl#update` is returning `true` wrongly though. From the location where it is called the comment indicates: `relayout when we have a breadcrumbs and when update changed its hidden-status`

But in reality you return `true` always even when the visibility does not change. Can you change that please?

https://github.com/microsoft/vscode/blob/475d3464e87ebea002b5046ddef8c94ff5eb6739/src/vs/workbench/browser/parts/editor/breadcrumbsControl.ts#L318-L318

**Comment by bpasero** [comment]
@jrieken thanks, I pushed another change to avoid repeated relayout calls from breadcrumbs, specifically this call is no longer needed...

https://github.com/microsoft/vscode/blob/61df1625efc575d7e8c90b76e1b747563568d48b/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L204

...because ever since tabs can wrap, there is logic in the layout call that will trigger a `reloayout` automatically once the height has changed from the last layout run:

https://github.com/microsoft/vscode/blob/eeff347cac7d1872bd090a0ce6f89e6a16b0bef2/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1374-L1374

And that should take care of breadcrumbs too.

I tested this briefly (jumping between normal text editor and untitled file) and couldn't find any issues, but maybe you could also give it a test.

The remaining layout calls seem to originate from the grid, so leaving up to Joao to decide wether we want / can improve that.

**Comment by jrieken** [comment]
> I tested this briefly (jumping between normal text editor and untitled file) and couldn't find any issues, but maybe you could also give it a test.

Cool. I will keep an eye open

**Comment by bpasero** [comment]
Not sure if related but doing some perf measurements on startup I see lots of time being spend on the critical path here:

![image](https://user-images.githubusercontent.com/900690/114025279-4029aa00-9875-11eb-9537-fde16845d675.png)

![image](https://user-images.githubusercontent.com/900690/114025310-4586f480-9875-11eb-81a5-c23d7f052cb7.png)

Possibly a forced layout somehow, not sure why the event would fire on startup though.

**Comment by joaomoreno** [comment]
Started looking at this again: what's **really** interesting is that if a file is open at the moment the workbench starts up, only a single `layout` call is made. 🤔

**Comment by jrieken** [comment]
@joaomoreno is this with tabs or no tabs?

**Comment by jrieken** [comment]
wait... stack trace tells me ;-) it's tabs...

**Comment by jrieken** [comment]
"fixed", at least breadcrumbs isn't triggering this eagerly anymore. Tho, when opening 1 file I count 6(**!**) layout calls and I'd say opening files is a more common scenario. 

@joaomoreno question: layout is about sizing and positioning child elements, right? So, by no means I support calling layout unnecessarily but isn't layout a good candidate for caching dimensions and then ignoring subsequent calls with the same dimensions?

**Comment by joaomoreno** [comment]
> @joaomoreno question: layout is about sizing and positioning child elements, right? So, by no means I support calling layout unnecessarily but isn't layout a good candidate for caching dimensions and then ignoring subsequent calls with the same dimensions?

Definitely, the question is where should that be? In the widgets themselves? In the widget components? In the workbench components? Everywhere? 🙈

**Comment by joaomoreno** [comment]
> "fixed", at least breadcrumbs isn't triggering this eagerly anymore. Tho, when opening 1 file I count 6(**!**) layout calls and I'd say opening files is a more common scenario.

We figured out in Slack:

- The scenario tested in this issue is: workbench startup
- Joh (and I) see 6 calls when opening a file by clicking it in the explorer, which is a separate scenario

**Comment by bpasero** [comment]
But my initial steps indicate the scenario, it is not startup, but opening a file:

```
Steps to Reproduce:

1. add a breakpoint or `console.log` into `editorGroupView.layout`
2. run vscode and have no file open
3. click on a file from the explorer
```

I see the following layout calls now from a fresh workspace when I open the first file:

```
Error
    at EditorGroupView.layout (editorGroupView.ts:1758)
    at LeafNode.layout (gridview.ts:781)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.relayout (splitview.ts:781)
    at SplitView.onViewChange (splitview.ts:616)
    at splitview.ts:692
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at event.ts:56
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at EditorControl.doSetActiveEditorPane (editorControl.ts:147)
    at EditorControl.doShowEditorPane (editorControl.ts:84)
    at EditorControl.openEditor (editorControl.ts:63)
    at editorGroupView.ts:1001
    at EditorGroupView.doShowEditor (editorGroupView.ts:1016)
    at EditorGroupView.doOpenEditor (editorGroupView.ts:982)
    at EditorGroupView.openEditor (editorGroupView.ts:910)
    at EditorService.openEditor (editorService.ts:629)
    at async DelegatingEditorService.editorOpenHandler (explorerViewlet.ts:224)

Error
    at EditorGroupView.layout (editorGroupView.ts:1758)
    at LeafNode.layout (gridview.ts:781)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.layout (splitview.ts:473)
    at BranchNode.layout (gridview.ts:380)
    at GridView.layout (gridview.ts:936)
    at SerializableGrid.layout (grid.ts:253)
    at SerializableGrid.layout (grid.ts:555)
    at GridWidgetView.layout (editorPart.ts:74)
    at CenteredViewLayout.layout (centeredViewLayout.ts:98)
    at EditorPart.doLayout (editorPart.ts:1096)
    at EditorPart.layout (editorPart.ts:1089)
    at LeafNode.layout (gridview.ts:781)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.relayout (splitview.ts:781)
    at SplitView.onViewChange (splitview.ts:616)
    at splitview.ts:692
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at event.ts:56
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at event.ts:91
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at event.ts:56
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at EditorControl.doSetActiveEditorPane (editorControl.ts:147)
    at EditorControl.doShowEditorPane (editorControl.ts:84)
    at EditorControl.openEditor (editorControl.ts:63)
    at editorGroupView.ts:1001
    at EditorGroupView.doShowEditor (editorGroupView.ts:1016)
    at EditorGroupView.doOpenEditor (editorGroupView.ts:982)
    at EditorGroupView.openEditor (editorGroupView.ts:910)
    at EditorService.openEditor (editorService.ts:629)
    at async DelegatingEditorService.editorOpenHandler (explorerViewlet.ts:224)

Error
    at EditorGroupView.layout (editorGroupView.ts:1758)
    at LeafNode.layout (gridview.ts:781)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.layout (splitview.ts:473)
    at BranchNode.layout (gridview.ts:380)
    at GridView.layout (gridview.ts:936)
    at SerializableGrid.layout (grid.ts:253)
    at SerializableGrid.layout (grid.ts:555)
    at GridWidgetView.layout (editorPart.ts:74)
    at CenteredViewLayout.layout (centeredViewLayout.ts:98)
    at EditorPart.doLayout (editorPart.ts:1096)
    at EditorPart.layout (editorPart.ts:1089)
    at LeafNode.layout (gridview.ts:781)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.layout (splitview.ts:473)
    at BranchNode.layout (gridview.ts:380)
    at HorizontalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.relayout (splitview.ts:781)
    at SplitView.onViewChange (splitview.ts:616)
    at splitview.ts:692
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at event.ts:91
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at event.ts:56
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at event.ts:91
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at event.ts:56
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at EditorControl.doSetActiveEditorPane (editorControl.ts:147)
    at EditorControl.doShowEditorPane (editorControl.ts:84)
    at EditorControl.openEditor (editorControl.ts:63)
    at editorGroupView.ts:1001
    at EditorGroupView.doShowEditor (editorGroupView.ts:1016)
    at EditorGroupView.doOpenEditor (editorGroupView.ts:982)
    at EditorGroupView.openEditor (editorGroupView.ts:910)
    at EditorService.openEditor (editorService.ts:629)
    at async DelegatingEditorService.editorOpenHandler (explorerViewlet.ts:224)

Error
    at EditorGroupView.layout (editorGroupView.ts:1758)
    at LeafNode.layout (gridview.ts:781)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.layout (splitview.ts:473)
    at BranchNode.layout (gridview.ts:380)
    at GridView.layout (gridview.ts:936)
    at SerializableGrid.layout (grid.ts:253)
    at SerializableGrid.layout (grid.ts:555)
    at GridWidgetView.layout (editorPart.ts:74)
    at CenteredViewLayout.layout (centeredViewLayout.ts:98)
    at EditorPart.doLayout (editorPart.ts:1096)
    at EditorPart.layout (editorPart.ts:1089)
    at LeafNode.layout (gridview.ts:781)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.layout (splitview.ts:473)
    at BranchNode.layout (gridview.ts:380)
    at HorizontalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.layout (splitview.ts:473)
    at BranchNode.layout (gridview.ts:380)
    at VerticalViewItem.layout (splitview.ts:138)
    at SplitView.layoutViews (splitview.ts:911)
    at SplitView.relayout (splitview.ts:781)
    at SplitView.onViewChange (splitview.ts:616)
    at splitview.ts:692
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at event.ts:91
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at event.ts:56
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at event.ts:56
    at event.ts:91
    at event.ts:91
    at Emitter.fire (event.ts:622)
    at event.ts:56
    at Emitter.fire (event.ts:622)
    at Emitter.fire (event.ts:624)
    at EditorControl.doSetActiveEditorPane (editorControl.ts:147)
    at EditorControl.doShowEditorPane (editorControl.ts:84)
    at EditorControl.openEditor (editorControl.ts:63)
    at editorGroupView.ts:1001
    at EditorGroupView.doShowEditor (editorGroupView.ts:1016)
    at EditorGroupView.doOpenEditor (editorGroupView.ts:982)
    at EditorGroupView.openEditor (editorGroupView.ts:910)
    at EditorService.openEditor (editorService.ts:629)
    at async DelegatingEditorService.editorOpenHandler (explorerViewlet.ts:224)

Error
    at EditorGroupView.layout (editorGroupView.ts:1758)
    at EditorGroupView.relayout (editorGroupView.ts:1776)
    at TabsTitleControl.doLayout (tabsTitleControl.ts:1371)
    at AnimationFrameQueueItem._runner (tabsTitleControl.ts:1332)
    at AnimationFrameQueueItem.execute (dom.ts:210)
    at animationFrameRunner (dom.ts:250)
```

That is 5 `layout` calls.

**Comment by joaomoreno** [comment]
Right, looking again.

**Comment by jrieken** [comment]
We should just assume that breadcrumbs are on - as they are by default- or read the config sooner. What I don't know is what we are laying out here. Isn't the title, breadcrumbs, and the editor-widget inside a single grid element and isn't that element already sized? What are we resizing/re-layouting here?

**Comment by joaomoreno** [comment]
> title control runs the layout but for perf reasons async (in a next animation frame)

Why is this? Entire complex widgets (splitview and list) layout in a sync fashion and a title bar doesn't? Can't you actually compute how much height you need instantly?

**Comment by bpasero** [comment]
It's a bit unfair that the async layout is also used for breadcrumbs because that is exactly why the relayout triggers: we only figure out that breadcrumbs are there when the animation frame executes, but we could run that part of the layout sync. I think a better way would be to only keep that code async where we know that DOM properties are accessed that can trigger a browser layout.

**Comment by jrieken** [comment]
> that it does which made me put the layout code into a "schedule at next animation frame"

Some FYI - running at next animation frame doesn't give you a guarantee that the DOM is pretty and that asking for layout info isn't expensive. It might well be that someone already modified the DOM so that an expensive re-compute is required but it is much less likely. I found this nice article a while back (https://developers.google.com/web/fundamentals/performance/rendering/avoid-large-complex-layouts-and-layout-thrashing) and then added `dom#measure` and `dom#modify`. However, those utils never became popular 😢

**Comment by bpasero** [comment]
@jrieken interesting, are you suggesting that I should be then using `dom#measure` in the layout call for tabs? I actually do not modify the dom in the layout method, I mainly measure.

**Comment by joaomoreno** [comment]
Fixed this all the way down to two layout calls: the actual one and the delayed one from @bpasero.

**Comment by bpasero** [comment]
This looks a lot lot better now, I verified that clicking on a file when no file is opened now only calls the `layout` method once and this originates from a `relayout` from the tabs title control itself. 

However, I went down the path trying to avoid the `relayout` by making the breadcrumbs layout sync, leaving the tabs layout async (which I still prefer given issues such as https://github.com/microsoft/vscode/issues/34745) and then in the end realised that even with a sync breadcrumbs layout, the `relayout` still happens. The reason is quite simple and I originally didn't think about it: when an editor opens, we actually do not go through the normal `layout` call from the editor group because no dimensions change (all we do is open an editor). However, when an editor opens with breadcrumbs, it will result in different dimensions for the title area compared to when no editor is open (`57px` vs `35px`) and as such a `relayout` is triggered to signal this to all controls that are layed out in the editor group.

But the good news is, that this call of `relayout` for the first editor is actually a no-op:
* since no editor is yet opened at that time, nothing happens in the editor control
* the re-entrant `layout` call to the tabs title control is also not doing anything because we are in a layout call

Since the `relayout` flow is still needed for the case when 1 editor is opened with breadcrumbs and you open a second editor without breadcrumbs (e.g. when opening an untitled file), I will stay away from making further changes here.
